internal/provider/resource_secret_access.go

```diff
@@ -0,0 +1,332 @@
+			ModelName:    state.Model.ValueString(),
+			SecretId:     state.SecretId.ValueString(),
+			Applications: applicationsInStateNotInPlan.([]string),
+		})
+		if err != nil {
+			resp.Diagnostics.AddError("Failed to revoke secret access", err.Error())
+			return
+		}
+	}
+
+	// grant access to applications that are in the plan but not in the state
+	if applicationsInPlanNotInState != nil {
+		err = s.client.Secrets.GrantSecretAccess(&juju.GrantRevokeSecretAccessInput{
+			ModelName:    state.Model.ValueString(),
+			SecretId:     state.SecretId.ValueString(),
+			Applications: applicationsInPlanNotInState.([]string),
+		})
+		if err != nil {
+			resp.Diagnostics.AddError("Failed to grant secret access", err.Error())
+			return
+		}
+	}
+
+	// Save updated data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+
+	s.trace(fmt.Sprintf("update secret access %q", state.SecretId))
+}
+
+func (s *secretAccessResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	// Prevent panic if the provider has not been configured.
+	if s.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "secret_access", "delete")
+		return
+	}
+
+	var state secretAccessResourceModel
+
+	// Read Terraform configuration state into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	applications := make([]string, len(state.Applications.Elements()))
+	resp.Diagnostics.Append(state.Applications.ElementsAs(ctx, &applications, false)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	err := s.client.Secrets.RevokeSecretAccess(&juju.GrantRevokeSecretAccessInput{
+		ModelName:    state.Model.ValueString(),
+		SecretId:     state.SecretId.ValueString(),
+		Applications: applications,
+	})
+	if err != nil {
+		resp.Diagnostics.AddError("Failed to revoke secret access", err.Error())
+		return
+	}
+
+	// Save empty list of applications into Terraform state
+	emptyApplicationList, errDiag := types.ListValue(types.StringType, []attr.Value{})
+	resp.Diagnostics.Append(errDiag...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	state.Applications = emptyApplicationList
+
+	// Save updated data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+
+	s.trace(fmt.Sprintf("revoke secret access %q", state.SecretId))
+}
+
+// difference returns the difference between two slices.
+func difference(slice1, slice2 interface{}) interface{} {
+	slice1Value := reflect.ValueOf(slice1)
+	slice2Value := reflect.ValueOf(slice2)
+
+	slice1Type := slice1Value.Type()
+	slice2Type := slice2Value.Type()
+
+	if slice1Type.Kind() != reflect.Slice || slice2Type.Kind() != reflect.Slice {
+		panic("Invalid data type. Expected slices.")
+	}
+
+	if slice1Type.Elem() != slice2Type.Elem() {
+		panic("Inconsistent slice types.")
+	}
+
+	set := make(map[interface{}]bool)
```

Comment:
does this need to be a map? I think the `if _, found := set[item]; found {` would still work if it was a `[]interface{}`.

Reply:
I've just rewrote this func to work with string slices